### DISCUSSION TITLE
How to use models::_entities::users in migration

### DISCUSSION THREAD

**Discussion by fan-tastic-z** [loco-rs/loco#392] [Q&A] [answered]
In migration, the data structure of the model cannot be used directly. How to configure Cargo.toml? I tried to add the configuration to the Cargo.toml of the migration, but it would cyclic package dependency error

in docs:
![image](https://github.com/loco-rs/loco/assets/142720801/c02f83b0-7a6e-4ee4-9486-962ae751291c)

**Comment by BWStearns** [comment]
I know this is a bit old but I am running into this as well.

**Reply by BWStearns** [reply]
@jondot is there any way we could show how to use the models in this example here? I keep running into the circular dependency issue that @fan-tastic-z did.

https://github.com/loco-rs/loco/blame/0fda51037c3461f1c74890cb0808515c51a227a6/docs-site/content/docs/the-app/models.md#L273

**Reply by jondot** [reply, marked as answer]
Thanks @BWStearns , after investigating this a bit, we're coming to conclusion that we're going to drop the recommendation to run data fixes in migrations using the actual entities.

The major reason is that while you're trying to modify data or DDL, your entities might already be incompatible (because you changed values or structure) and so it makes less sense to use the `model` abstractions here.

The original suggestion came from the SeaORM docs (https://www.sea-ql.org/SeaORM/docs/migration/writing-migration/#tip-3-seed-data-with-entity) however, we think that since there's some manual work involved in that example, the expectation is from the user to "operate" what ever issue that arises. In Loco we want things to happen automatically when booting up.

To conclude, we'll remove the recommendation to use the models in this instance, and recommend using raw SQL if must, or create custom tasks that do data fixes per user requirement, where you can use your high level models.

Thanks for bringing this up again!

**Reply by BWStearns** [reply]
Thanks for looking into it! When I get around to doing my data migration I'll take a look at using tasks and if I find a good pattern I'll put in a PR with an example so we can point to that in the docs.

**Reply by BWStearns** [reply]
```
        let _ = manager
            .get_connection()
            .execute_unprepared("DELETE FROM table_being_migrated WHERE user_id IS NULL")
            .await;
```

For anyone else looking and not super familiar with sqlx/seaorm, I ended up going with something like the above. Seems to work.

**Reply by Phundrak** [reply]
I’ve had success with simply using strings:

```rust
m.create_index(
    sea_query::Index::create()
        .name("idx_users_username")
        .table("users")
        .col("username")
        .to_owned(),
).await?;
```

This also seems to work for foreign keys:
```rust
.foreign_key(
    ForeignKey::create()
        .name("fk-users-user_id-to-profiles")
        .from("profiles", "user_id")
        .to("users", "id")
)
```

**Comment by kevinclancy** [comment]
This also applies to the documentation for creating indices, dropping columns, and adding columns, doesn't it? I'm making an index creation migration, and the documentation still recommends that I use the entities.

I'm going to just use raw SQL, but the documentation for these things should probably be fixed.

**Reply by kaplanelad** [reply]
Thanks @kevinclancy. Do you want to open a PR to fix the docs?

**Reply by kevinclancy** [reply]
Sure. I've created a PR [here](https://github.com/loco-rs/loco/pull/1549/files).
